src/Checkout/ExpressCheckout/Service/ExpressShippingCallbackService.php

```diff
@@ -0,0 +1,146 @@
+use Shopware\Core\Framework\Validation\DataBag\RequestDataBag;
+use Shopware\Core\System\Country\CountryEntity;
+use Shopware\Core\System\SalesChannel\Context\AbstractSalesChannelContextFactory;
+use Shopware\Core\System\SalesChannel\Context\SalesChannelContextService;
+use Shopware\Core\System\SalesChannel\SalesChannel\AbstractContextSwitchRoute;
+use Shopware\Core\System\SalesChannel\SalesChannelContext;
+use Shopware\PayPalSDK\Struct\V2\Order\PurchaseUnit;
+use Swag\PayPal\Checkout\CheckoutException;
+use Swag\PayPal\OrdersApi\Builder\Util\AmountProvider;
+
+#[Package('checkout')]
+class ExpressShippingCallbackService
+{
+    /**
+     * @internal
+     */
+    public function __construct(
+        private readonly CartService $cartService,
+        private readonly EntityRepository $countryRepository,
+        private readonly AmountProvider $amountProvider,
+        private readonly AbstractContextSwitchRoute $contextSwitchRoute,
+        private readonly AbstractSalesChannelContextFactory $salesChannelContextFactory,
+        private readonly LoggerInterface $logger,
+    ) {
+    }
+
+    /**
+     * @param array<string, mixed> $shippingAddress
+     *
+     * @return array<int, array<string, mixed>>
+     */
+    public function recalculateCart(
+        string $paypalOrderId,
+        array $shippingAddress,
+        SalesChannelContext $salesChannelContext,
+    ): array {
+        $this->logger->debug('Shipping callback: Recalculating cart for address change', [
+            'paypalOrderId' => $paypalOrderId,
+            'countryCode' => $shippingAddress['country_code'] ?? null,
+            'fullAddress' => $shippingAddress,
+        ]);
+
+        // Get country by ISO code
+        $countryCode = $shippingAddress['country_code'] ?? null;
+        if (!$countryCode) {
+            $this->logger->error('Shipping callback: Missing country code in shipping address');
+            throw CheckoutException::expressMissingCountryCode();
+        }
+
+        $country = $this->getCountryByIso($countryCode, $salesChannelContext->getContext());
+        if (!$country) {
+            $this->logger->error('Country not found', ['countryCode' => $countryCode]);
+            throw CheckoutException::expressCountryNotFound($countryCode);
+        }
+
+        // Since a new customer was logged in, the context changed in the system,
+        // but this doesn't effect the current context given as parameter.
+        // Because of that a new context for the cart recalculation is created
+        $this->logger->debug('Switching context to new country', ['countryId' => $country->getId()]);
+        $this->contextSwitchRoute->switchContext(
+            new RequestDataBag([
+                SalesChannelContextService::COUNTRY_ID => $country->getId(),
+            ]),
+            $salesChannelContext
+        );
+
+        $newSalesChannelContext = $this->salesChannelContextFactory->create(
+            $salesChannelContext->getToken(),
```

Comment:
Although the default implementation always returns the same token, a different one may be returned:
```suggestion
        $newToken = $this->contextSwitchRoute->switchContext(
            new RequestDataBag([
                SalesChannelContextService::COUNTRY_ID => $country->getId(),
            ]),
            $salesChannelContext
        )->getToken();

        $newSalesChannelContext = $this->salesChannelContextFactory->create(
            $newToken,
```

Reply:
Would you mind telling me how does that make a difference? What happen if I keep the old token vs change to a new? 
I'm just curious, it's hard to look at the code and figure that out 🤔

Reply:
Is it if I use the same token, the Shopware cart context would also be affected with the changed country in Paypal UI?

Reply:
The reason for this is that it is technically possible for another token to return, rather than this actually happening. Like all routes, the `ContextSwitchRoute` extends `AbstractContextSwitchRoute` and dispatches several events - allowing for context manipulations or completely decorating or overriding the business logic. Only one thing is for sure, you get a token response back, which will load the context you're expecting :smile:

Reply:
Okay that's a lot :)) thank you, I'll have to analyze the code versus your comment to undertstand better :))